File: screener/backtester/metrics.py
"""Backtest performance metrics.

All metrics derive from the portfolio equity curve and the aligned benchmark.
Alpha/beta use a simple OLS fit via ``numpy.polyfit`` — no sklearn.
"""

from __future__ import annotations

import math
from typing import Iterable

import numpy as np
import pandas as pd

from screener.backtester.models import Trade
from screener.regime import classify_regimes


TRADING_DAYS_PER_YEAR = 252
_EULER_MASCHERONI = 0.5772156649015329


def _daily_returns(equity: pd.Series) -> pd.Series:
    if equity.empty or len(equity) < 2:
        return pd.Series(dtype=float)
    return equity.pct_change().dropna()


def _cagr(equity: pd.Series) -> float:
    if equity.empty or len(equity) < 2:
        return 0.0
    start = float(equity.iloc[0])
    end = float(equity.iloc[-1])
    if start <= 0:
        return 0.0
    years = max(len(equity) / TRADING_DAYS_PER_YEAR, 1e-9)
    return (end / start) ** (1.0 / years) - 1.0


def _max_drawdown(equity: pd.Series) -> float:
    if equity.empty:
        return 0.0
    peak = equity.cummax()
    dd = (equity - peak) / peak
    return float(dd.min()) if not dd.empty else 0.0


def _sharpe(daily: pd.Series, rf: float = 0.0) -> float:
    if daily.empty or daily.std(ddof=0) == 0:
        return 0.0
    excess = daily - rf / TRADING_DAYS_PER_YEAR
    return float(excess.mean() / excess.std(ddof=0) * np.sqrt(TRADING_DAYS_PER_YEAR))


def _vol_annual(daily: pd.Series) -> float:
    if daily.empty:
        return 0.0
    return float(daily.std(ddof=0) * np.sqrt(TRADING_DAYS_PER_YEAR))


def _alpha_beta(daily: pd.Series, bench_daily: pd.Series) -> tuple[float, float]:
    if daily.empty or bench_daily.empty:
        return 0.0, 0.0
    aligned = pd.concat([daily, bench_daily], axis=1).dropna()
    if len(aligned) < 2:
        return 0.0, 0.0
    x = aligned.iloc[:, 1].to_numpy()
    y = aligned.iloc[:, 0].to_numpy()
    if x.std() == 0:
        return 0.0, 0.0
    slope, intercept = np.polyfit(x, y, 1)
    # annualize alpha (intercept is per-day)
    return float(intercept * TRADING_DAYS_PER_YEAR), float(slope)


def _exposure(
    equity_index: pd.DatetimeIndex, trades: Iterable[Trade], slot_count: int
) -> float:
    trades = list(trades)
    if not trades or len(equity_index) == 0:
        return 0.0
    open_count = pd.Series(0, index=equity_index, dtype=int)
    for t in trades:
        entry = pd.Timestamp(t.entry_date)
        exit_ = pd.Timestamp(t.exit_date)
        mask = (equity_index >= entry) & (equity_index <= exit_)
        open_count.loc[mask] += 1
    return float(open_count.mean() / max(slot_count, 1))


def _sortino(daily: pd.Series, rf: float = 0.0) -> float:
    if daily.empty:
        return 0.0
    excess = daily - rf / TRADING_DAYS_PER_YEAR
    downside = excess[excess < 0]
    if downside.empty or downside.std(ddof=0) == 0:
        return 0.0
    return float(excess.mean() / downside.std(ddof=0) * np.sqrt(TRADING_DAYS_PER_YEAR))


def _calmar(equity: pd.Series) -> float:
    if equity.empty or len(equity) < 2:
        return 0.0
    mdd = _max_drawdown(equity)
    if mdd >= 0:
        return 0.0
    return float(_cagr(equity) / abs(mdd))


def _phi(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def _phi_inv(p: float) -> float:
    """Standard-normal inverse CDF via bisection on erf. Good to ~1e-12."""
    if p <= 0.0:
        return -float("inf")
    if p >= 1.0:
        return float("inf")
    lo, hi = -8.0, 8.0
    for _ in range(80):
        mid = 0.5 * (lo + hi)
        if _phi(mid) < p:
            lo = mid
        else:
            hi = mid
    return 0.5 * (lo + hi)


def _psr(daily: pd.Series, sr_benchmark_annual: float = 0.0) -> float:
    """Probabilistic Sharpe Ratio (López de Prado, 2012).

    Probability that the *true* annualized Sharpe exceeds ``sr_benchmark_annual``,
    corrected for sample size and non-normality (skew, excess kurtosis). Returns
    a value in [0, 1].
    """
    if daily.empty or len(daily) < 30:
        return 0.0
    T = len(daily)
    sr_per = _sharpe(daily) / math.sqrt(TRADING_DAYS_PER_YEAR)
    sr_bench_per = sr_benchmark_annual / math.sqrt(TRADING_DAYS_PER_YEAR)
    skew = float(daily.skew()) if daily.std(ddof=0) else 0.0
    kurt_excess = float(daily.kurt()) if daily.std(ddof=0) else 0.0
    denom_sq = 1.0 - skew * sr_per + (kurt_excess / 4.0) * sr_per * sr_per
    denom = math.sqrt(max(denom_sq, 1e-12))
    z = (sr_per - sr_bench_per) * math.sqrt(max(T - 1, 1)) / denom
    return _phi(z)


def _dsr(
    daily: pd.Series,
    n_trials: int = 1,
    sr_trial_std_annual: float = 0.5,
) -> float:
    """Deflated Sharpe Ratio (López de Prado, 2014).

    Like PSR, but the benchmark Sharpe is the *expected maximum* across
    ``n_trials`` independent strategies under the null — i.e. the bar a random
    strategy would clear just from multiple-testing luck. ``sr_trial_std_annual``
    is the cross-trial std of annualized Sharpes (0.5 is a reasonable default
    for equity strategies); pass the measured value if you have it.
    """
    if n_trials <= 1:
        return _psr(daily, 0.0)
    sr0_annual = sr_trial_std_annual * (
        (1.0 - _EULER_MASCHERONI) * _phi_inv(1.0 - 1.0 / n_trials)
        + _EULER_MASCHERONI * _phi_inv(1.0 - 1.0 / (n_trials * math.e))
    )
    return _psr(daily, sr_benchmark_annual=sr0_annual)


def _invested_return(trades: Iterable[Trade]) -> float:
    """Capital-deployed-only total return.

    Ignores idle cash in the denominator: sums realized PnL across all closed
    trades and divides by total capital that actually touched the market
    (sum of entry_cost). Exposes the dead-cash gap even when the engine's
    reinvestment path is on — a low ratio of equity_return to invested_return
    indicates a large share of capital sat idle.
    """
    trades = list(trades)
    total_cost = sum(float(t.entry_cost) for t in trades)
    total_pnl = sum(float(t.pnl) for t in trades)
    if total_cost <= 0:
        return 0.0
    return total_pnl / total_cost


def compute_regime_metrics(benchmark: pd.Series, trades: list[Trade]) -> dict:
    """Per-regime trade stats keyed by each trade's entry-date regime.

    The regime is classified from the (point-in-time) benchmark close curve;
    each trade is bucketed by the benchmark regime in force on its entry date
    (most recent benchmark date at or before entry). Returns keys of the form
    ``regime_<label>_trades`` / ``regime_<label>_win_rate`` /
    ``regime_<label>_avg_return`` for each label with at least one trade.
    """
    if benchmark is None or benchmark.empty or not trades:
        return {}
    regimes = classify_regimes(benchmark)
    idx = regimes.index
    by_label: dict[str, list[float]] = {}
    for t in trades:
        pos = int(idx.searchsorted(pd.Timestamp(t.entry_date), side="right")) - 1
        label = str(regimes.iloc[pos]) if pos >= 0 else "unknown"
        by_label.setdefault(label, []).append(float(t.return_pct))
    out: dict = {}
    for label, rets in by_label.items():
        out[f"regime_{label}_trades"] = len(rets)
        out[f"regime_{label}_win_rate"] = sum(1 for r in rets if r > 0) / len(rets)
        out[f"regime_{label}_avg_return"] = sum(rets) / len(rets)
    return out


def compute_metrics(
    equity: pd.Series,
    benchmark: pd.Series,
    trades: list[Trade],
    slot_count: int,
    n_trials: int = 1,
) -> dict:
    daily = _daily_returns(equity)
    bench_daily = (
        _daily_returns(benchmark) if not benchmark.empty else pd.Series(dtype=float)
    )
    total_return = (
        float(equity.iloc[-1] / equity.iloc[0] - 1.0)
        if len(equity) >= 2 and equity.iloc[0] > 0
        else 0.0
    )
    alpha, beta = _alpha_beta(daily, bench_daily)
    hit_rate = (
        float(sum(1 for t in trades if t.pnl > 0) / len(trades)) if trades else 0.0
    )
    bench_return = (
        float(benchmark.iloc[-1] / benchmark.iloc[0] - 1.0)
        if len(benchmark) >= 2 and benchmark.iloc[0] > 0
        else 0.0
    )
    return {
        "total_return": total_return,
        "cagr": _cagr(equity),
        "vol_annual": _vol_annual(daily),
        "sharpe": _sharpe(daily),
        "sortino": _sortino(daily),
        "calmar": _calmar(equity),
        "psr": _psr(daily, sr_benchmark_annual=0.0),
        "dsr": _dsr(daily, n_trials=n_trials),
        "max_drawdown": _max_drawdown(equity),
        "hit_rate": hit_rate,
        "alpha_annual": alpha,
        "beta": beta,
        "exposure": _exposure(equity.index, trades, slot_count),
        "benchmark_return": bench_return,
        "trade_count": len(trades),
        "invested_return": _invested_return(trades),
    }

File: screener-rs/src/backtester/metrics.rs
use crate::backtester::models::Trade;
use chrono::NaiveDate;
use std::collections::BTreeMap;

const TRADING_DAYS_PER_YEAR: f64 = 252.0;
const EULER_MASCHERONI: f64 = 0.577_215_664_901_532_9;

fn daily_returns(curve: &[(NaiveDate, f64)]) -> Vec<f64> {
    curve
        .windows(2)
        .filter_map(|pair| {
            let prev = pair[0].1;
            let curr = pair[1].1;
            if prev.is_finite() && curr.is_finite() && prev != 0.0 {
                Some(curr / prev - 1.0)
            } else {
                None
            }
        })
        .collect()
}

fn mean(values: &[f64]) -> f64 {
    if values.is_empty() {
        0.0
    } else {
        values.iter().sum::<f64>() / values.len() as f64
    }
}

fn std_pop(values: &[f64]) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    let avg = mean(values);
    let var = values.iter().map(|v| (v - avg).powi(2)).sum::<f64>() / values.len() as f64;
    var.sqrt()
}

fn cagr(curve: &[(NaiveDate, f64)]) -> f64 {
    if curve.len() < 2 || curve[0].1 <= 0.0 {
        return 0.0;
    }
    let years = (curve.len() as f64 / TRADING_DAYS_PER_YEAR).max(1e-9);
    (curve.last().unwrap().1 / curve[0].1).powf(1.0 / years) - 1.0
}

fn max_drawdown(curve: &[(NaiveDate, f64)]) -> f64 {
    if curve.is_empty() {
        return 0.0;
    }
    let mut peak = curve[0].1;
    let mut worst = 0.0;
    for (_, equity) in curve {
        if *equity > peak {
            peak = *equity;
        }
        if peak != 0.0 {
            let dd = (*equity - peak) / peak;
            if dd < worst {
                worst = dd;
            }
        }
    }
    worst
}

fn sharpe(daily: &[f64], rf: f64) -> f64 {
    let sd = std_pop(daily);
    if daily.is_empty() || sd == 0.0 {
        return 0.0;
    }
    let excess: Vec<f64> = daily
        .iter()
        .map(|r| r - rf / TRADING_DAYS_PER_YEAR)
        .collect();
    mean(&excess) / std_pop(&excess) * TRADING_DAYS_PER_YEAR.sqrt()
}

fn vol_annual(daily: &[f64]) -> f64 {
    std_pop(daily) * TRADING_DAYS_PER_YEAR.sqrt()
}

fn sortino(daily: &[f64], rf: f64) -> f64 {
    if daily.is_empty() {
        return 0.0;
    }
    let excess: Vec<f64> = daily
        .iter()
        .map(|r| r - rf / TRADING_DAYS_PER_YEAR)
        .collect();
    let downside: Vec<f64> = excess.iter().copied().filter(|r| *r < 0.0).collect();
    let downside_std = std_pop(&downside);
    if downside.is_empty() || downside_std == 0.0 {
        return 0.0;
    }
    mean(&excess) / downside_std * TRADING_DAYS_PER_YEAR.sqrt()
}

fn calmar(curve: &[(NaiveDate, f64)]) -> f64 {
    let mdd = max_drawdown(curve);
    if curve.len() < 2 || mdd >= 0.0 {
        return 0.0;
    }
    cagr(curve) / mdd.abs()
}

fn alpha_beta(equity_daily: &[f64], benchmark_daily: &[f64]) -> (f64, f64) {
    let n = equity_daily.len().min(benchmark_daily.len());
    if n < 2 {
        return (0.0, 0.0);
    }
    let x = &benchmark_daily[..n];
    let y = &equity_daily[..n];
    let mx = mean(x);
    let my = mean(y);
    let var_x = x.iter().map(|v| (v - mx).powi(2)).sum::<f64>() / n as f64;
    if var_x == 0.0 {
        return (0.0, 0.0);
    }
    let cov = x
        .iter()
        .zip(y)
        .map(|(a, b)| (a - mx) * (b - my))
        .sum::<f64>()
        / n as f64;
    let beta = cov / var_x;
    let intercept = my - beta * mx;
    (intercept * TRADING_DAYS_PER_YEAR, beta)
}

fn exposure(equity: &[(NaiveDate, f64)], trades: &[Trade], slot_count: usize) -> f64 {
    if trades.is_empty() || equity.is_empty() {
        return 0.0;
    }
    let total_open: usize = equity
        .iter()
        .map(|(day, _)| {
            trades
                .iter()
                .filter(|trade| *day >= trade.entry_date && *day <= trade.exit_date)
                .count()
        })
        .sum();
    total_open as f64 / equity.len() as f64 / slot_count.max(1) as f64
}

fn phi(x: f64) -> f64 {
    0.5 * (1.0 + erf(x / 2.0_f64.sqrt()))
}

fn erf(x: f64) -> f64 {
    // Abramowitz and Stegun 7.1.26; sufficient for PSR/DSR gating metrics.
    let sign = if x < 0.0 { -1.0 } else { 1.0 };
    let x = x.abs();
    let t = 1.0 / (1.0 + 0.3275911 * x);
    let a1 = 0.254829592;
    let a2 = -0.284496736;
    let a3 = 1.421413741;
    let a4 = -1.453152027;
    let a5 = 1.061405429;
    let y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * (-x * x).exp();
    sign * y
}

fn phi_inv(p: f64) -> f64 {
    if p <= 0.0 {
        return f64::NEG_INFINITY;
    }
    if p >= 1.0 {
        return f64::INFINITY;
    }
    let (mut lo, mut hi) = (-8.0, 8.0);
    for _ in 0..80 {
        let mid = 0.5 * (lo + hi);
        if phi(mid) < p {
            lo = mid;
        } else {
            hi = mid;
        }
    }
    0.5 * (lo + hi)
}

fn skew(values: &[f64]) -> f64 {
    let sd = std_pop(values);
    if values.is_empty() || sd == 0.0 {
        return 0.0;
    }
    let avg = mean(values);
    values.iter().map(|v| ((v - avg) / sd).powi(3)).sum::<f64>() / values.len() as f64
}

fn kurt_excess(values: &[f64]) -> f64 {
    let sd = std_pop(values);
    if values.is_empty() || sd == 0.0 {
        return 0.0;
    }
    let avg = mean(values);
    values.iter().map(|v| ((v - avg) / sd).powi(4)).sum::<f64>() / values.len() as f64 - 3.0
}

fn psr(daily: &[f64], sr_benchmark_annual: f64) -> f64 {
    if daily.len() < 30 {
        return 0.0;
    }
    let sr_per = sharpe(daily, 0.0) / TRADING_DAYS_PER_YEAR.sqrt();
    let sr_bench_per = sr_benchmark_annual / TRADING_DAYS_PER_YEAR.sqrt();
    let denom_sq = 1.0 - skew(daily) * sr_per + (kurt_excess(daily) / 4.0) * sr_per.powi(2);
    let denom = denom_sq.max(1e-12).sqrt();
    phi((sr_per - sr_bench_per) * ((daily.len() - 1) as f64).sqrt() / denom)
}

fn dsr(daily: &[f64], n_trials: usize, sr_trial_std_annual: f64) -> f64 {
    if n_trials <= 1 {
        return psr(daily, 0.0);
    }
    let n = n_trials as f64;
    let sr0_annual = sr_trial_std_annual
        * ((1.0 - EULER_MASCHERONI) * phi_inv(1.0 - 1.0 / n)
            + EULER_MASCHERONI * phi_inv(1.0 - 1.0 / (n * std::f64::consts::E)));
    psr(daily, sr0_annual)
}

fn invested_return(trades: &[Trade]) -> f64 {
    let total_cost: f64 = trades.iter().map(|trade| trade.entry_cost).sum();
    let total_pnl: f64 = trades.iter().map(|trade| trade.pnl).sum();
    if total_cost <= 0.0 {
        0.0
    } else {
        total_pnl / total_cost
    }
}

pub fn compute_metrics(
    equity: &[(NaiveDate, f64)],
    benchmark: &[(NaiveDate, f64)],
    trades: &[Trade],
    slot_count: usize,
    n_trials: usize,
) -> BTreeMap<String, f64> {
    let daily = daily_returns(equity);
    let bench_daily = daily_returns(benchmark);
    let total_return = if equity.len() >= 2 && equity[0].1 > 0.0 {
        equity.last().unwrap().1 / equity[0].1 - 1.0
    } else {
        0.0
    };
    let benchmark_return = if benchmark.len() >= 2 && benchmark[0].1 > 0.0 {
        benchmark.last().unwrap().1 / benchmark[0].1 - 1.0
    } else {
        0.0
    };
    let (alpha, beta) = alpha_beta(&daily, &bench_daily);
    let hit_rate = if trades.is_empty() {
        0.0
    } else {
        trades.iter().filter(|trade| trade.pnl > 0.0).count() as f64 / trades.len() as f64
    };
    BTreeMap::from([
        ("total_return".to_string(), total_return),
        ("cagr".to_string(), cagr(equity)),
        ("vol_annual".to_string(), vol_annual(&daily)),
        ("sharpe".to_string(), sharpe(&daily, 0.0)),
        ("sortino".to_string(), sortino(&daily, 0.0)),
        ("calmar".to_string(), calmar(equity)),
        ("psr".to_string(), psr(&daily, 0.0)),
        ("dsr".to_string(), dsr(&daily, n_trials, 0.5)),
        ("max_drawdown".to_string(), max_drawdown(equity)),
        ("hit_rate".to_string(), hit_rate),
        ("alpha_annual".to_string(), alpha),
        ("beta".to_string(), beta),
        ("exposure".to_string(), exposure(equity, trades, slot_count)),
        ("benchmark_return".to_string(), benchmark_return),
        ("trade_count".to_string(), trades.len() as f64),
        ("invested_return".to_string(), invested_return(trades)),
    ])
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;

    fn d(day: u32) -> NaiveDate {
        NaiveDate::from_ymd_opt(2024, 1, day).unwrap()
    }

    #[test]
    fn max_drawdown_metric_matches_python_case() {
        let eq = vec![(d(1), 100.0), (d(2), 110.0), (d(3), 105.0), (d(4), 120.0)];
        let metrics = compute_metrics(&eq, &eq, &[], 1, 1);
        assert_relative_eq!(metrics["max_drawdown"], -5.0 / 110.0);
    }
}
